slides/advanced-03-iterative.qmd

```diff
@@ -0,0 +1,658 @@
+
+How can we use the outputs to choose the next point to measure?
+
+<br> 
+
+[_Acquisition functions_](https://ekamperi.github.io/machine%20learning/2021/06/11/acquisition-functions.html) take the predicted mean and variance and use them to balance: 
+
+ - _exploration_:  new candidates should explore new areas.
+ - _exploitation_: new candidates must stay near existing values. 
+
+Exploration focuses on the variance, exploitation is about the mean. 
+
+## Acquisition Functions
+
+We'll use an acquisition function to select a new candidate.
+
+The most popular method appears to be _expected improvement_ ([EI](https://arxiv.org/pdf/1911.12809.pdf)) above the current best results. 
+ 
+  - Zero at existing data points. 
+  - The _expected_ improvement is integrated overall all possible improvement ("expected" in the probability sense). 
+
+We would probably pick the point with the largest EI as the next point. 
+
+(there are other functions beyond EI)
+
+## Expected Improvement
+
+```{r}
+#| label: gp-iter-0-ei
+#| echo: false
+#| out-width: 50%
+#| fig-align: center
+#| fig-width: 6
+#| fig-height: 4.25
+
+scaled_obj %>%
+  filter(.iter == 0) %>% 
+  ggplot(aes(learn_rate, scaled)) +
+  geom_point(data = bayes_points %>% filter(.iter == 0), aes(y = zero)) +
+  geom_line(alpha = 1 / 2, col = "#D95F02", linewidth = 1) +
+  scale_x_log10()  +
+  labs(y = "Expected Improvement", x = "Learning Rate")
+```
+
+## Iteration
+
+Once we pick the candidate point, we measure performance for it (e.g. resampling). 
+
+Another GP is fit, EI is recomputed, and so on. 
+
+We stop when we have completed the allowed number of iterations _or_ if we don't see any improvement after a pre-set number of attempts. 
+
+
+## GP Fit with four points
+
+```{r}
+#| label: gp-iter-1
+#| echo: false
+#| out-width: 50%
+#| fig-align: center
+#| fig-width: 6
+#| fig-height: 4.25
+
+scaled_obj %>%
+  filter(.iter == 1) %>% 
+  ggplot(aes(learn_rate, .mean)) +
+  geom_line(data = grid_points, aes(y = mean), alpha = 1 / 4, 
+            col = "#0b84a5", linewidth = 1.5) +
+  geom_line() +
+  geom_point(data = bayes_points %>% filter(.iter == 1), aes(y = mean)) +
+  geom_ribbon(aes(ymin = .mean - .sd, ymax = .mean + .sd),
+              alpha = 1 / 7) +
+  scale_x_log10() +
+  labs(y = "MAE", x = "Learning Rate")
+```
+
+
+## Expected Improvement
+
+```{r}
+#| label: gp-iter-1-ei
+#| echo: false
+#| out-width: 50%
+#| fig-align: center
+#| fig-width: 6
+#| fig-height: 4.25
+
+scaled_obj %>%
+  filter(.iter == 1) %>% 
+  ggplot(aes(learn_rate, scaled)) +
+  geom_point(data = bayes_points %>% filter(.iter == 1), aes(y = zero)) +
+  geom_line(alpha = 1 / 2, col = "#D95F02", linewidth = 1) +
+  scale_x_log10()  +
+  labs(y = "Expected Improvement", x = "Learning Rate")
+```
+
+
+## GP Evolution
+
+
+```{r}
+#| label: gp-anime
+#| echo: false
+#| out-width: 50%
+#| fig-align: center
+#| fig-width: 6
+#| fig-height: 4.25
+
+knitr::include_graphics("animations/anime_gp.gif")
+```
```

Comment:
I don't see any content on this or the next slide when I look at the rendered html. Do others have this issue?

Reply:
On a separate computer, I fetched the PR and compiled it. It is there for me. Is the EI gif on the next slide working?

Reply:
So strange--maybe another oddity from my differing revealjs install. Will leave this for after NYR👍